### The Survival Opportunity on Titanic
Titanic disaster happened over 100 years ago, the Titanic sank after colliding with an iceberg, killing 1502 out of 2224 passengers and crew. There were some safety problems associate with the tragedy that led to the big loss of life, one of them was not enough lifeboats for everyone, people were panic and suffered in cold, but different group of people on board also have different chances on surviving. According to the analysis, women, children and upper class had more chances to survive than others. 

Here is a brief Machine Learning model analysis and survive prediction on the on boarding passengers. We will split the passenger data into two groups, one is used to train the model, and another one will be used to test the model. The target is the result that the passenger 'survived' or 'not survived', and we will use K Nearest Neighbor, Decision Tree, Support Vector Machine, Logistic Regression to model the data, and compare the accuracy of the models.  

In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv("titanic.csv")
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38      1   
2                             Heikkinen, Miss. Laina  female  26      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35      1   
4                           Allen, Mr. William Henry    male  35      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500     ?        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250     ?        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500     ?        S

In [3]:
df.shape

(891, 12)

In [4]:
df["Survived"].value_counts()

0    549
1    342
Name: Survived, dtype: int64

In [5]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age             object
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [6]:
df.describe(include = "all")

PassengerId    Survived      Pclass                     Name   Sex  \
count    891.000000  891.000000  891.000000                      891   891   
unique          NaN         NaN         NaN                      891     2   
top             NaN         NaN         NaN  Braund, Mr. Owen Harris  male   
freq            NaN         NaN         NaN                        1   577   
mean     446.000000    0.383838    2.308642                      NaN   NaN   
std      257.353842    0.486592    0.836071                      NaN   NaN   
min        1.000000    0.000000    1.000000                      NaN   NaN   
25%      223.500000    0.000000    2.000000                      NaN   NaN   
50%      446.000000    0.000000    3.000000                      NaN   NaN   
75%      668.500000    1.000000    3.000000                      NaN   NaN   
max      891.000000    1.000000    3.000000                      NaN   NaN   

        Age       SibSp       Parch  Ticket        Fare Cabin Embarked  
count   891  891.000000  891.000000     891  891.000000   891      891  
unique   89         NaN         NaN     681         NaN   148        4  
top       ?         NaN         NaN  347082         NaN     ?        S  
freq    177         NaN         NaN       7         NaN   687      644  
mean    NaN    0.523008    0.381594     NaN   32.204208   NaN      NaN  
std     NaN    1.102743    0.806057     NaN   49.693429   NaN      NaN  
min     NaN    0.000000    0.000000     NaN    0.000000   NaN      NaN  
25%     NaN    0.000000    0.000000     NaN    7.910400   NaN      NaN  
50%     NaN    0.000000    0.000000     NaN   14.454200   NaN      NaN  
75%     NaN    1.000000    0.000000     NaN   31.000000   NaN      NaN  
max     NaN    8.000000    6.000000     NaN  512.329200   NaN      NaN

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            891 non-null object
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          891 non-null object
Embarked       891 non-null object
dtypes: float64(1), int64(5), object(6)
memory usage: 83.7+ KB


In [8]:
df.replace("?", np.NaN, inplace=True)
df.drop(["PassengerId", "Name", "Ticket", "Cabin"], axis=1, inplace=True)
avg_age = df["Age"].astype("float").mean(axis=0)
print("Average of age:", avg_age)

Average of age: 29.69911764705882


In [9]:
df["Age"].replace(np.nan, avg_age, inplace=True)
df["Sex"].replace(to_replace=["male", "female"],value= [0, 1], inplace = True)
df["Embarked"].replace(to_replace=["C", "Q", "S"],value= [1, 2, 3], inplace = True)
df.dropna(subset=["Embarked"], axis=0, inplace=True)
# reset index, because we droped two rows
df.reset_index(drop=True, inplace=True)

In [10]:
Feature =df[["Pclass", "Sex", "Age", "SibSp", "Parch", "Fare","Embarked"]]
y=df["Survived"]

In [11]:
df.head()

Survived  Pclass  Sex Age  SibSp  Parch     Fare  Embarked
0         0       3    0  22      1      0   7.2500       3.0
1         1       1    1  38      1      0  71.2833       1.0
2         1       3    1  26      0      0   7.9250       3.0
3         1       1    1  35      1      0  53.1000       3.0
4         0       3    0  35      0      0   8.0500       3.0

In [12]:
X = Feature
X[0:5]

Pclass  Sex Age  SibSp  Parch     Fare  Embarked
0       3    0  22      1      0   7.2500       3.0
1       1    1  38      1      0  71.2833       1.0
2       3    1  26      0      0   7.9250       3.0
3       1    1  35      1      0  53.1000       3.0
4       3    0  35      0      0   8.0500       3.0

In [13]:
from sklearn import preprocessing
X = preprocessing.StandardScaler().fit(X).transform(X)
X[0:5]

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/sklearn/preprocessing/data.py:625: DataConversionWarning: Data with input dtype int64, float64, object were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/ipykernel_launcher.py:2: DataConversionWarning: Data with input dtype int64, float64, object were all converted to float64 by StandardScaler.
  


array([[ 0.82520863, -0.73534203, -0.59049493,  0.43135024, -0.47432585,
        -0.50023975,  0.58683958],
       [-1.57221121,  1.35991138,  0.64397101,  0.43135024, -0.47432585,
         0.78894661, -1.93955453],
       [ 0.82520863,  1.35991138, -0.28187844, -0.47519908, -0.47432585,
        -0.48664993,  0.58683958],
       [-1.57221121,  1.35991138,  0.41250865,  0.43135024, -0.47432585,
         0.42286111,  0.58683958],
       [ 0.82520863, -0.73534203,  0.41250865, -0.47519908, -0.47432585,
        -0.4841333 ,  0.58683958]])

In [14]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=4)
print ('Train set:', X_train.shape,  y_train.shape)
print ('Test set:', X_test.shape,  y_test.shape)

Train set: (711, 7) (711,)
Test set: (178, 7) (178,)


# K Nearest Neighbor(KNN)

In [15]:
from sklearn.neighbors import KNeighborsClassifier
k = 3
#Train Model and Predict  
kNN_model = KNeighborsClassifier(n_neighbors=k).fit(X_train,y_train)
kNN_model

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=None, n_neighbors=3, p=2,
           weights='uniform')

In [16]:
# just for sanity chaeck
yhat = kNN_model.predict(X_test)
yhat[0:5]

array([0, 1, 0, 1, 1])

In [17]:
# Best k
from sklearn.neighbors import KNeighborsClassifier
Ks=15
mean_acc=np.zeros((Ks-1))
std_acc=np.zeros((Ks-1))
for n in range(1,Ks):
     
    kNN_model = KNeighborsClassifier(n_neighbors=n).fit(X_train,y_train)
    yhat = kNN_model.predict(X_test)
    mean_acc[n-1]=np.mean(yhat==y_test);
    
    std_acc[n-1]=np.std(yhat==y_test)/np.sqrt(yhat.shape[0])
print(mean_acc)
print(' ')
print(n-2, mean_acc.max()) 

[0.79775281 0.76966292 0.78651685 0.76966292 0.76404494 0.78089888
 0.79213483 0.78651685 0.7752809  0.78089888 0.78651685 0.80337079
 0.78651685 0.78651685]
 
12 0.8033707865168539


# Decision Tree

In [18]:
from sklearn.tree import DecisionTreeClassifier
DT_model = DecisionTreeClassifier(criterion="entropy", max_depth = 4)
DT_model.fit(X_train,y_train)
DT_model

DecisionTreeClassifier(class_weight=None, criterion='entropy', max_depth=4,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [19]:
yhat = DT_model.predict(X_test)
yhat

array([0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1,
       0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0,
       1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0,
       1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0,
       0, 0])

# Support Vector Machine

In [20]:
from sklearn import svm
SVM_model = svm.SVC()
SVM_model.fit(X_train, y_train) 

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/sklearn/svm/base.py:196: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto_deprecated',
  kernel='rbf', max_iter=-1, probability=False, random_state=None,
  shrinking=True, tol=0.001, verbose=False)

In [21]:
yhat = SVM_model.predict(X_test)
yhat

array([0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1,
       0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0,
       1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0,
       1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0,
       0, 0])

# Logistic Regression

In [22]:
from sklearn.linear_model import LogisticRegression
LR_model = LogisticRegression(C=0.01).fit(X_train,y_train)
LR_model

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=0.01, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=None, solver='warn',
          tol=0.0001, verbose=0, warm_start=False)

In [23]:
yhat = LR_model.predict(X_test)
yhat

array([1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0,
       1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 0,
       0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0,
       1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0,
       1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0,
       0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1,
       0, 0])

# Model Evaluation using Test set

In [24]:
from sklearn.metrics import jaccard_similarity_score
from sklearn.metrics import f1_score
from sklearn.metrics import log_loss

###  Evaluation 

In [25]:
knn_yhat = kNN_model.predict(X_test)
print("KNN Jaccard index: %.2f" % jaccard_similarity_score(y_test, knn_yhat))
print("KNN F1-score: %.2f" % f1_score(y_test, knn_yhat, average='weighted') )

KNN Jaccard index: 0.79
KNN F1-score: 0.78


In [26]:
DT_yhat = DT_model.predict(X_test)
print("DT Jaccard index: %.2f" % jaccard_similarity_score(y_test, DT_yhat))
print("DT F1-score: %.2f" % f1_score(y_test, DT_yhat, average='weighted') )

DT Jaccard index: 0.78
DT F1-score: 0.76


In [27]:
SVM_yhat = SVM_model.predict(X_test)
print("SVM Jaccard index: %.2f" % jaccard_similarity_score(y_test, SVM_yhat))
print("SVM F1-score: %.2f" % f1_score(y_test, SVM_yhat, average='weighted') )

SVM Jaccard index: 0.82
SVM F1-score: 0.81


In [28]:
LR_yhat = LR_model.predict(X_test)
LR_yhat_prob = LR_model.predict_proba(X_test)
print("LR Jaccard index: %.2f" % jaccard_similarity_score(y_test, LR_yhat))
print("LR F1-score: %.2f" % f1_score(y_test, LR_yhat, average='weighted') )
print("LR LogLoss: %.2f" % log_loss(y_test, LR_yhat_prob))

LR Jaccard index: 0.75
LR F1-score: 0.75
LR LogLoss: 0.52


### Conclusion
The project just includes part of the dataset, and we built different models on the data. After data wrangling, like replace the missing values in the age with the mean, dropped rows in the 'Embarked' variable, get dummy variables for categorical variable, we split the cleaned dataset, and build different models to predict the target variables. The accuracy table of using different evaluation metrics is as below:

| Algorithm          | Jaccard | F1-score | LogLoss |
|--------------------|---------|----------|---------|
| KNN                | 0.79    | 0.78     | NA      |
| Decision Tree      | 0.78    | 0.76     | NA      |
| SVM                | 0.82    | 0.81     | NA      |
| LogisticRegression | 0.75    | 0.75     | 0.52    |